index.js

```diff
@@ -36,73 +48,49 @@ function cacheSet(key, value) {
  * @return {String} a hash representing the stats of this module and all its descendents
  */
 module.exports = function hashForDep(name, dir, _hashTreeOverride, _skipCache) {
-  var skipCache = false;
-  var key, hash;
+  var skipCache = (typeof _hashTreeOverride === 'function' || _skipCache === true) ? true : false;
 
-  if (typeof _hashTreeOverride === 'function' || _skipCache === true) {
-    skipCache = true;
-  } else {
-    key = cacheKey(name, dir);
-  }
+  var caches = skipCache ? new CacheGroup() : PROCESS_CACHE;
+
+  var hashTreeFn = (_hashTreeOverride || hashTree);
+
+  var nameDirKey = genNameDirKey(name, dir);
+  var realPathKey, realPath;
+  var hash;
 
   var heimdallNodeOptions = {
     name: 'hashForDep(' + name + ')',
     hashForDep: true,
     dependencyName: name,
     rootDir: dir,
     skipCache: skipCache,
-    cacheKey: key
+    cacheKey: nameDirKey
   };
 
-  var heimdallNode = heimdall.start(heimdallNodeOptions, HashForDepSchema);
-
-  if (skipCache === false && CACHE.has(key)) {
-    logger.info('cache hit: %s', key);
-    hash = CACHE.get(key);
-  } else {
-    var start = Date.now();
-
-    var inputHashes = statPathsFor(name, dir).map(function(statPath) {
-      var hashFn = _hashTreeOverride || hashTree;
-
-      heimdallNode.stats.paths++;
-
-      var key = 'hashOf:' + statPath;
-      var hash;
-
-      if (skipCache === false && CACHE.has(key)) {
-        logger.debug('HIT', key)
-        hash = CACHE.get(key);
-      } else {
-        logger.debug('MISS', key)
-        hash = hashFn(statPath);
-        if (skipCache === false) {
-          CACHE.set(key, hash);
-        }
-      }
-      return hash;
-    }).join(0x00);
-
-    hash = crypto.createHash('sha1').
-      update(inputHashes).digest('hex');
+  var h = heimdall.start(heimdallNodeOptions, HashForDepSchema);
 
-    logger.info('cache miss: %s, paths: %d, took: %dms', key, heimdallNode.stats.paths, Date.now() - start);
-
-    if (skipCache === false) {
-      CACHE.set(key, hash);
-    }
+  try {
+    return ModuleEntry.locate(caches, name, dir, hashTreeFn).getHash(h);
```

Comment:
I've done a pass through the code to fix comments and do the co-location of those three functions (only one is now public, the resolvePackagePath one, but the others are private, named with _, and exported for testing. I've verified that that all seems to work, and done a live debug to see that the right stuff is called at the right places. I'll add the new tests and resolve windows issues (if there are any) tomorrow after the conf. at Microsoft. At the worst I should have something for you Friday morning. I hesitate to check things in right now just in case they're somehow bad, though I see no evidence of that right now.